src/prover_disk.hpp

```diff
@@ -238,10 +253,78 @@ class DiskProver {
     std::string filename;
     std::vector<uint8_t> memo;
     std::vector<uint8_t> id;  // Unique plot id
-    uint8_t k;
+    uint8_t k{0};
     std::vector<uint64_t> table_begin_pointers;
     std::vector<uint64_t> C2;
 
+    void Read(std::vector<uint8_t>& id_in, std::vector<uint8_t>& memo_in, uint8_t& k_in, std::vector<uint64_t>& table_begin_pointers_in, std::vector<uint64_t>& C2_in) const
+    {
+        std::lock_guard<std::mutex> lock(_mtx);
+
+        std::ifstream disk_file(filename, std::ios::in | std::ios::binary);
+
+        if (!disk_file.is_open()) {
+            throw std::invalid_argument("Invalid file " + filename);
+        }
+        struct plot_header header{};
+        // 19 bytes  - "Proof of Space Plot" (utf-8)
+        // 32 bytes  - unique plot id
+        // 1 byte    - k
+        // 2 bytes   - format description length
+        // x bytes   - format description
+        // 2 bytes   - memo length
+        // x bytes   - memo
+
+        SafeRead(disk_file, (uint8_t*)&header, sizeof(header));
+        if (memcmp(header.magic, "Proof of Space Plot", sizeof(header.magic)) != 0)
+            throw std::invalid_argument("Invalid plot header magic");
+
+        uint16_t fmt_desc_len = Util::TwoBytesToInt(header.fmt_desc_len);
+
+        if (fmt_desc_len == kFormatDescription.size() &&
+        !memcmp(header.fmt_desc, kFormatDescription.c_str(), fmt_desc_len)) {
+            // OK
+        } else {
+            throw std::invalid_argument("Invalid plot file format");
+        }
+        id_in = std::vector<uint8_t>(kIdLen);
+        memcpy(id_in.data(), header.id, sizeof(header.id));
+        k_in = header.k;
+        SafeSeek(disk_file, offsetof(struct plot_header, fmt_desc) + fmt_desc_len);
+
+        uint8_t size_buf[2];
+        SafeRead(disk_file, size_buf, 2);
+        memo_in.resize(Util::TwoBytesToInt(size_buf));
+        SafeRead(disk_file, memo_in.data(), memo_in.size());
+
+        table_begin_pointers_in = std::vector<uint64_t>(nTableBeginPointerCount, 0);
+        C2_in = std::vector<uint64_t>();
```

Comment:
The only benefit I can think of, of passing these vectors in as out-parameters is that they can reuse memory they may already have allocated. However, when you assign them like this, rather than `.resize(0)`, you defeat that on benefit.

I would think, though, that these allocations are cheap anyway, and that it would make sense to return them instead of taking them as out-parameters.

Reply:
I didn't use the `resize` approach here because i wanted to save one line :smile: (setting the contents or at least the first element of `table_begin_pointers_out` to zero) and it obviously doesn't really matter given the size of this vectors.